sycl/test-e2e/WorkGroupMemory/test-plan.md

```diff
@@ -0,0 +1,148 @@
+that all APIs are consistent with respect to other APIs.
+
+#### Runtime diagnostics on misuse
+
+`work_group_memory` objects can only be used in `parallel_for(nd_range)`
+kernels, so the test should ensure that we throw a synchronious `exception`
+with the `errc::kernel_argument` error code if `work_group_memory` was used by
+other types of kernels (`single_task` and `parallel_for(range)`).
+
+#### Consistency of address-taking operations
+
+The test need to check that `operator&()` and `get_multi_ptr().get_raw()` both
+return the same address
+
+#### Copy constructors and assignment operator
+
+The test need to check that `operator&()` and `get_multi_ptr().get_raw()` both
+return correct address after `work_group_memory` was copy-constructed, or
+copy-assigned from another `work_group_memory` object.
+
+#### Operations on different types
+
+This test should check that an object of `work_group_memory` class behaves like
+an object of a type `work_group_memory` is templated on, i.e. all operations
+possible on an underlying type can also be performed on `work_group_memory`
+object and produce correct results.
+
+#### Number of kernel arguments produced by `work_group_memory` objects
+
+One of the main reasons behind the extension is to provide a lightweight
+alternative to `local_accessor`. To make sure that the implementation is indeed
+lightweight, there should be a test which ensures that every `work_group_memory`
+object results in a single kernel argument generated by the compiler.
+
+### End-to-end tests
+
+Tests in this category perform some meaningful actions with the extension to
+see that the extension works in a scenarios which mimic real-life usage of the
+extension.
+
+#### Basic usage
+
+This test is intended to check that memory provided through `work_group_memory`
+is indeed shared between work-items within a work-group and can be properly
+accessed for reading-writing to it.
+
+A test should allocate an array of integers through `work_group_memory` and
+submit a single (for simplicity) work-group where every work-item would store
+an element from a global buffer into that allocated local array which
+corresponds to its local id.
+
+Then, a single work-item should perform a reduction of that local array, i.e.
+compute sum of its elements. The result should be communicated back to host and
+verified there.
+
+Some of the operations performed on `work_group_memory` object should be done in
+a separate helper function to check that we can pass `work_group_memory` objects
+to functions and correctcly access data they reference there.
+
+#### Use of multiple `work_group_memory` objects
+
+This test is similar to the one above, but a kernel should operate on more than
+one `work_group_memory` object. For example, the second `work_group_memory`
+object could be used to broadcast reduction value to all work-items in a
+work-group.
+
+#### Use of `work_group_memory` with `atomic_ref`
```

Comment:
@Pennycook Is`work_group_memory` supposed to be used with `atomic_ref` as underlying data? 
From what I've gathered, `atomic_ref` does not have a default constructor and since the work group memory interface does not allow the user to supply one, what would be the intended semantics in this case?
Do we internally allocate some memory and point the atomic_ref to it?
If so, would that memory itself be local or global memory?
Currently, my implementation produces incorrect results and seemingly so does `local_accessor` based on my testing.
Or perhaps, I'm just misusing it.
Here is what I ran:
```
#include <sycl/sycl.hpp>

using namespace sycl;

int main() {
  queue Q;
  Q.submit([&](handler &cgh) {
     local_accessor<
         atomic_ref<int, memory_order::seq_cst, memory_scope::work_group>, 0>
         lacc;
     cgh.parallel_for(nd_range{{1}, {1}}, [=](nd_item<> it) {
       atomic_ref<int, memory_order::seq_cst, memory_scope::work_group> &ref = lacc{cgh};
       ref.store(25);
       assert(ref.load() == 25); // assertion fails
     });
   }).wait();
}
```
Same result when using work group memory:
```
#include <sycl/sycl.hpp>

using namespace sycl;

int main() {
  queue Q;
  Q.submit([&](handler &cgh) {
     ext::oneapi::experimental::work_group_memory<
         atomic_ref<int, memory_order::seq_cst, memory_scope::work_group>>
         wgm{cgh};
     cgh.parallel_for(nd_range{{1}, {1}}, [=](nd_item<> it) {
       atomic_ref<int, memory_order::seq_cst, memory_scope::work_group> &ref =
           wgm;
       ref.store(25);
       assert(ref.load() == 25);
     });
   }).wait();
}
```
Also tagging @gmlueck for awareness.

Reply:
This is not how `atomic_ref` works.  You're not supposed to create an array of `atomic_ref` objects, but to create an `atomic_ref` that references the object you would like to modify atomically:

```c++
#include <sycl/sycl.hpp>

using namespace sycl;

int main() {
  queue Q;
  Q.submit([&](handler &cgh) {
     local_accessor<int, 0> lacc{cgh};
     cgh.parallel_for(nd_range{{1}, {1}}, [=](nd_item<> it) {
       atomic_ref<int, memory_order::seq_cst, memory_scope::work_group> ref(lacc);
       ref.store(25);
       assert(ref.load() == 25); // assertion fails
     });
   }).wait();
}
```

Reply:
Thanks for the clarificcation!